### PR TITLE
[WIP] Refine analysis w.r.t. DifferentiableGraphs, fuse separately

### PR DESCRIPTION
**Pull request by t-vi**
Currently, if we have a graph with some requires_grad inputs, we create DifferentiableGraphs whereever we can, even if there are no gradient-needing tensors in it. This PR changes this to remove
DifferentiableGraphs with no grad-requiring members.
At the moment, we only try to fuse things in the differentiable graphs. Afer this PR, we instead check that we don't involve gradient-requiring tensors outside DifferentiableGraphs.

This simplifies the graphs a bit when you have only parts requiring gradients. But it will be even more useful when we fuse (non-grad) inplace ops.

I need to double-check that I really didn't forget to add checks for gradient-requiring tensors and add tests.

### CODE REVIEW

**Comment by t-vi** [comment]
This broke again, apparently related to quantization (but that might be just the symptom or even just the tests).
@ngimel commented on some effects of this PR on slack, but unless this seems useful to anyone, I'd be inclined to drop the PR.

**Comment by t-vi** [comment]
I don't want to sit on this. I still think it would be a good idea to have this refined analysis, but I won't get to it anytime soon.
